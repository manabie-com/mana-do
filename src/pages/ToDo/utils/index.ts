import { Todo, TodoStatus } from '../../../models/todo';
const created_date = new Date().toISOString();
export const defaultToDos = {
  content: 'Content',
  created_date,
  status: TodoStatus.ACTIVE,
  id: 'testId',
  user_id: 'firstUser',
} as Todo;

export function spawnToDos(count: number): Todo[] {
  return new Array(count).fill(defaultToDos).map(todo => {
    todo.id = `testId-${Math.random()}`;
    return todo;
  });
}
