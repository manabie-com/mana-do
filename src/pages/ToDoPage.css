.ToDo__container {
    border: 1px solid rgba(0, 0, 0, 0.13);
    border-radius: 8px;
    min-width: 400px;
    width: 40vw;
    height: 750px;
    margin-top: 5vh;
    margin-bottom: 5vh;
    padding: 24px;
    box-shadow: 2px 2px 1px rgba(0, 0, 0, 0.09), 3px 2px 3px rgba(0, 0, 0, 0.05);
    background: #a5a5a5;
    position: relative;
}

.Todo__creation {
    display: flex;
}

.Todo__input {
    background-color: #d8d8d8;
    flex: 1 1;
}

.Todo__input::placeholder {
    color: rgb(75, 75, 75);
}

.ToDo__list {
    display: flex;
    flex-direction: column;
    margin-top: 1.5rem;
    padding-right: 15px;
    height: 630px;
    overflow: auto;
    border-bottom: 2px solid rgba(41, 39, 39, 0.323);
}

.Todo__content {
    flex: 1 1;
}

.Todo__action,
.Todo__delete {
    width: 24px;
    height: 24px;
    flex-shrink: 0;
}

.Todo__toolbar {
    position: absolute;
    bottom: 20px;
    min-width: 400px;
    width: 40vw;
    display: flex;
    justify-content: space-between;
}

.Todo__tabs {
    display: flex;
    justify-content: center;
}

.Todo__tabs > *:not(:last-child) {
    margin-right: 8px;
}
