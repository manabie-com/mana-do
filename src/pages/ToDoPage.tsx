import React, {useEffect, useReducer, useState} from 'react';
import reducer, {initialState} from 'store/reducer';
import {setTodos} from 'store/actions';
import {TodoStatus} from 'models/todo';
import TodoToolBar from "components/TodoToolBar";
import TodoList from "components/TodoList";
import Service from 'service';
import TodoCreation from "components/TodoCreation";

type EnhanceTodoStatus = TodoStatus | 'ALL';

const ToDoPage = () => {
    const [{todos}, dispatch] = useReducer(reducer, initialState);
    const [showing, setShowing] = useState<EnhanceTodoStatus>('ALL');

    useEffect(() => {
        (async () => {
            const resp = await Service.getTodos();
            dispatch(setTodos(resp || []));
        })()
    }, [])

    useEffect(() => {
        (async () => {
            await Service.updateTodos(todos);
        })()
    }, [todos])

    const propsTodoList = {showing, setShowing, dispatch, todos};
    const propsTodoToolBar = {showing, setShowing, dispatch, todos};

    return (
        <div className="Todo__container">
            <TodoCreation dispatch={dispatch} />
            <TodoList {...propsTodoList} />
            <TodoToolBar {...propsTodoToolBar} />
        </div>
    );
};

export default ToDoPage;
