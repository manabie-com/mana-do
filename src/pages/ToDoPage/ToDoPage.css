h1.title {
	font-size: 120px;
    text-align: center;
    color: darkseagreen;
    padding: 2rem;
}

.ToDo__container {
	display: flex;
	width: 100% !important;
	justify-content: center;
	flex-direction: column;
    background: white;
	box-shadow: 0 2px 4px 0 rgb(0 0 0 / 20%), 0 25px 50px 0 rgb(0 0 0 / 10%);
    flex: 0 0 50%;
    max-width: 50%;
}

.ToDo__list {
	display: flex;
	flex-direction: column;
}

.ToDo__empty-msg {
	min-height: 40px;
	display: flex;
	align-items: center;
	padding: 0 30px;
	font-style: italic;
}

@media (max-width: 1500px) {
	.ToDo__container {
		flex: 0 0 60%;
		max-width: 60%;
	}
}

@media (max-width: 1100px) {
	.ToDo__container {
		flex: 0 0 70%;
		max-width: 70%;
	}
}

@media (max-width: 900px) {
	.ToDo__container {
		flex: 0 0 95%;
		max-width: 95%;
	}
}