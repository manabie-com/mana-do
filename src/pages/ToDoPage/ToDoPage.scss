@import "../../assets/scss/variables";

.to-do {
  display: flex;
  flex-direction: column;
  align-items: center;
  &__main {
    width: 100%;
    max-width: 600px;
    margin-bottom: 24px;
  }
  &__container {
    padding: 0 16px;
    width: 100%;
    display: flex;
    flex-direction: column;
    align-items: center;
  }
}
