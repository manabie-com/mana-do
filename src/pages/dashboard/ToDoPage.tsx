import React, { useContext, useEffect, useRef, useState } from 'react';
import { RouteComponentProps } from 'react-router-dom';
import {
    setTodos,
    createTodo,
    deleteAllTodos,
    actSignOut,
} from '../../store/actions';
import Service from '../../service';
import { TodoStatus } from '../../models/todo';
import { AppContext } from '../../store/context';
import AppButton from '../../commons/AppButton';
import AppTitle from '../../components/AppTitle';
import TodoList from '../../components/TodoList';
import AppRow from '../../commons/AppRow';
import AppCol from '../../commons/AppCol';
import AppText from '../../commons/AppText';
import AppInput from '../../commons/AppInput';
import { useNotAuthenticated } from '../../hooks/useNotAuthenticated';
import AppFooter from '../../components/AppFooter';

// type EnhanceTodoStatus = TodoStatus | 'ALL';


const ToDoPage = ({ history }: RouteComponentProps) => {
    useNotAuthenticated();
    const { state, dispatch } = useContext(AppContext);
    const { auth } = state;
    const inputRef = useRef<HTMLInputElement>(null);
    const [activeTab, setActiveTab] = useState(localStorage.getItem('filter')||'ALL');
    useEffect(() => {
        (async () => {
            try {
                const resp = await Service.getTodos();
                dispatch(setTodos(resp.data || []));

            } catch (error) {
                if (error.status === 401) {
                    history.push('/')
                }
            }
        })()
    }, [dispatch, history])

    const onCreateTodo = async (e: React.KeyboardEvent<HTMLInputElement>) => {
        if (e.key === 'Enter' && inputRef.current) {

            try {
                const resp = await Service.createTodo(inputRef.current.value);
                dispatch(createTodo(resp.data));
                inputRef.current.value = '';
            } catch (e) {
                if (e.status === 401) {
                    history.push('/')
                }
            }
        }
    }

    const onDeleteAllTodo = async () => {
        dispatch(deleteAllTodos());
        try {
            const response = await Service.clearTodos();
            console.log(response);
        } catch (error) {
            if (error.status === 401) {
                history.push('/');
            }
        }
    }

    const filterButtonsJSX = ():React.ReactNode => {
        return (<>
        <AppButton btnType={activeTab==='ALL'?'primary':'default'}  key="ALL" onClick={()=>{setActiveTab('ALL'); localStorage.setItem('filter','ALL')}}>All</AppButton>
        {Object.entries(TodoStatus).map(item=><AppButton key={item[0]} btnType={activeTab===item[1]?'primary':'default'} onClick={(e:React.MouseEventHandler<HTMLButtonElement>)=>{setActiveTab(item[1]); localStorage.setItem('filter',item[1])}}>{item[0]}</AppButton>)}
        </>)
    }

    const onClickSignOut = () => {
        dispatch(actSignOut());
        localStorage.removeItem('token');
        history.push('/');
    }

    return (
        <AppRow className="justify-content-center">
            <AppCol xs={5} sm ={9} md={10} lg={8} xl={8}>
            <div className="ToDo__container">
                <AppTitle />
                <p>Hi, <AppText strong>{auth.currentUser?.firstName || ''}</AppText><br /><AppButton btnType="link" href="#" onClick={onClickSignOut}>Sign out</AppButton></p>
                <div className="Todo__creation">
                    <AppInput
                        ref={inputRef}
                        className="Todo__input"
                        placeholder="What need to be done?"
                        onKeyDown={onCreateTodo}
                        style={{marginBottom: 20}}
                    />
                </div>                
                    <TodoList filter={activeTab}></TodoList>
                <div className="Todo__toolbar">
                    <div className="Todo__tabs">
                        {filterButtonsJSX()}
                    </div>
                    <AppButton btnType="primary" onClick={onDeleteAllTodo}>
                        Clear all
                    </AppButton>
                </div>
            </div>
            <AppFooter textColor="white" />
            </AppCol>
        </AppRow>
    );
};

export default ToDoPage;