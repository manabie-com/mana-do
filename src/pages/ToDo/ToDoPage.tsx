import React from 'react';
import ToDoItem from './ToDoItem';
import ToDoInput from './ToDoInput';
import Service from '../../service';
import ToDoToolbar from './ToDoToolbar';
import { setTodos } from '../../store/actions';
import { AppContext } from '../../context/toDoContext';
import { TransitionGroup } from 'react-transition-group';
import { saveToLocalStorage } from '../../utils/localStorage';
import { AppState } from '../../store/reducer';

export default function ToDoPage() {
  const { state, dispatch } = React.useContext(AppContext);

  React.useEffect(() => {
    (async () => {
      const resp = await Service.getTodos();
      dispatch(setTodos(resp || []));
    })();
  }, [dispatch]);

  React.useEffect(() => {
    saveToLocalStorage(state.todos);
  }, [state.todos]);

  return (
    <div className='ToDo__container'>
      <ToDoInput />
      <FilteredToDoList state={state} />
      <ToDoToolbar />
    </div>
  );
}

function FilteredToDoList({ state }: { state: AppState }) {
  const { todos, filter } = state;
  if (filter === 'ALL') {
    return (
      <div className='ToDo__list'>
        <TransitionGroup>
          {todos.map(todo => (
            <ToDoItem key={todo.id} todo={todo} />
          ))}
        </TransitionGroup>
      </div>
    );
  }

  const filteredTodos = todos.filter(todo => todo.status === filter);
  return (
    <div className='ToDo__list'>
      <TransitionGroup>
        {filteredTodos.map(todo => (
          <ToDoItem key={todo.id} todo={todo} />
        ))}
      </TransitionGroup>
    </div>
  );
}
