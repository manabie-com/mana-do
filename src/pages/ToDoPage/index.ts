import ToDoPage from "./ToDoPage";

export default ToDoPage;