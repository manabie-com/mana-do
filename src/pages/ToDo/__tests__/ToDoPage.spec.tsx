import React from 'react';
import ToDoPage from '../ToDoPage';
import { config } from 'react-transition-group';
import * as storage from '../../../utils/localStorage';
import { AppProvider } from '../../../context/toDoContext';
import { fireEvent, render, screen } from '@testing-library/react';
import { spawnToDos } from '../utils';
import userEvent from '@testing-library/user-event';

// Disabled waits in Transistion
config.disabled = true;

jest.mock('@mojs/core');
jest.mock('../../../utils/mojs', () => ({
  __esModule: true,
  playCheckedEffect: jest.fn().mockImplementation(() => null),
}));

let spySaveToLocalStorage = jest.spyOn(storage, 'saveToLocalStorage');
let spyGetToDosFromLocalStorage = jest.spyOn(
  storage,
  'getToDosFromLocalStorage',
);

afterEach(() => {
  spySaveToLocalStorage.mockReset();
  spySaveToLocalStorage.mockRestore();
  spyGetToDosFromLocalStorage.mockReset();
  spyGetToDosFromLocalStorage.mockRestore();
});

describe('ToDoPage', () => {
  it('Should match snapshot', () => {
    const { container } = render(<ToDoPage />);
    expect(container).toMatchSnapshot();
  });
});

describe('Render ToDoPage', () => {
  it('Should have default todo on first load', async () => {
    const todos = spawnToDos(1);
    spyGetToDosFromLocalStorage.mockReturnValue(todos);
    const { findAllByTestId } = render(<ToDoPage />, { wrapper: AppProvider });
    const toDoItems = await findAllByTestId(/todo-item/i);

    expect(toDoItems).toHaveLength(todos.length);
  });
});

describe('Create todos', () => {
  it('Should allow to create three todos', async () => {
    const user = userEvent.setup();
    render(<ToDoPage />, { wrapper: AppProvider });

    const toDos = ['item1', 'item2', 'item3'];
    for (let i = 0; i < toDos.length; i++) {
      await user.keyboard(`${toDos[i]}{Enter}`);
      expect(
        await screen.findByText(toDos[i], { selector: 'span' }),
      ).toBeInTheDocument();
    }

    const todoItems = await screen.findAllByText(/item/i, { selector: 'span' });
    expect(todoItems).toHaveLength(toDos.length);
  });

  it('Should save new todos to localStorage', async () => {
    const user = userEvent.setup();
    render(<ToDoPage />, { wrapper: AppProvider });
    const todoInput = await screen.findByPlaceholderText(
      /what need to be done?/i,
    );
    todoInput.focus();
    await user.keyboard('test persistance to-dos{Enter}');
    const todoItems = await screen.findByText(/test persistance to-dos/i, {
      selector: 'span',
    });

    expect(todoItems).toBeInTheDocument();
  });
});

describe('Edit todo', () => {
  it('Should allow edit todo by double click on it', async () => {
    const { findByText } = render(<ToDoPage />, { wrapper: AppProvider });
    const defaultToDo = await findByText(/content/i, {
      selector: 'span',
    });

    // await user.dblClick(defaultToDo); not woking
    // https://stackoverflow.com/questions/60495903/testing-react-contenteditable-with-react-testing-library
    // Workaround: https://github.com/testing-library/dom-testing-library/pull/235
    // Mimic user typing action
    fireEvent.blur(defaultToDo, { target: { textContent: 'edit content' } });
    defaultToDo.dispatchEvent(
      new KeyboardEvent('keydown', {
        bubbles: true,
        key: 'Enter',
      }),
    );

    const updateTodo = await screen.findByText(/edit content/i, {
      selector: 'span',
    });

    expect(updateTodo).toBeInTheDocument();
  });

  it('Should discard change when clicking outside', async () => {
    const { findByText } = render(<ToDoPage />, { wrapper: AppProvider });

    const defaultTodo = await findByText(/content/i, {
      selector: 'span',
    });

    // Change todo content and click outside
    fireEvent.blur(defaultTodo, { target: { textContent: 'discard edit' } });

    expect(spySaveToLocalStorage).not.toHaveBeenCalled();
  });
});

describe('Active/complete todo', () => {
  it('Should allow users to click on checkbox items they have completed', async () => {
    const user = userEvent.setup();
    spyGetToDosFromLocalStorage.mockReturnValue(spawnToDos(2));
    render(<ToDoPage />, { wrapper: AppProvider });
    const todoCheckboxs = await screen.findAllByLabelText('todo-item-checkbox');
    await user.click(todoCheckboxs[0]);

    expect(todoCheckboxs[0]).toBeChecked();
  });
});

describe('Delete todo', () => {
  it('Should allow users to delete a single todo', async () => {
    const user = userEvent.setup();
    render(<ToDoPage />, { wrapper: AppProvider });

    const deleteButtons = await screen.findAllByLabelText('delete-todo-button');
    await user.click(deleteButtons[0]);

    expect(deleteButtons[0]).not.toBeInTheDocument();
  });
});

describe('ToDoToolbar', () => {
  it('Should allow users to toggle all todos', async () => {
    const user = userEvent.setup();
    const { findByLabelText, findAllByLabelText } = render(<ToDoPage />, {
      wrapper: AppProvider,
    });

    const toggleAllToDosCheckbox = await findByLabelText('toggle-all-todos');
    const individualsToDoCheckboxes = await findAllByLabelText(
      'todo-item-checkbox',
    );
    await user.click(toggleAllToDosCheckbox);

    expect(toggleAllToDosCheckbox).toBeChecked();
    individualsToDoCheckboxes.forEach(checkbox => {
      expect(checkbox).toBeChecked();
    });
  });

  it('Should allow users to clear all todos', async () => {
    const user = userEvent.setup();
    render(<ToDoPage />, { wrapper: AppProvider });
    const clearAllToDosButton = await screen.findByText(/Clear all todos/i);
    await user.click(clearAllToDosButton);

    expect(() => screen.getAllByLabelText(/todo-item-checkbox/i)).toThrow(
      'Unable to find a label with the text of: /todo-item-checkbox/i',
    );
  });

  it('Should be able to select filter selection', async () => {
    const user = userEvent.setup();
    render(<ToDoPage />);
    await user.selectOptions(screen.getByRole('combobox'), 'All');
    const allOption = screen.getByRole('option', {
      name: 'All',
    }) as HTMLOptionElement;

    expect(allOption.selected).toBe(true);

    await user.selectOptions(screen.getByRole('combobox'), 'Active');
    const activeOption = screen.getByRole('option', {
      name: /active/i,
    }) as HTMLOptionElement;

    expect(activeOption.selected).toBe(true);

    await user.selectOptions(screen.getByRole('combobox'), 'Completed');
    const completedOption = screen.getByRole('option', {
      name: /completed/i,
    }) as HTMLOptionElement;

    expect(completedOption.selected).toBe(true);
  });
});
