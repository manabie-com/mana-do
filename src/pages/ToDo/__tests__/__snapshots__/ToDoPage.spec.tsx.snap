// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoPage Should match snapshot 1`] = `
<div>
  <div
    class="ToDo__container"
  >
    <div
      class="Todo__creation"
    >
      <input
        class="Todo__input"
        placeholder="What need to be done?"
      />
    </div>
    <div
      class="ToDo__list"
    >
      <div />
    </div>
    <div
      class="Todo__toolbar"
    >
      <div />
      <div
        class="Todo__tabs"
      >
        <label
          for="filter"
        >
          View
        </label>
        <select
          id="filter"
          name="filter"
        >
          <option
            selected=""
            value="ALL"
          >
            All
          </option>
          <option
            value="ACTIVE"
          >
            Active
          </option>
          <option
            value="COMPLETED"
          >
            Completed
          </option>
        </select>
      </div>
      <button
        arial-label="delete-all-todos"
        class="Action__btn"
        disabled=""
      >
        Clear all todos
      </button>
    </div>
  </div>
</div>
`;
